devito/passes/iet/misc.py

```diff
@@ -226,23 +225,26 @@ def _(expr, langbb):
 
 @_lower_macro_math.register(Max)
 @_lower_macro_math.register(sympy.Max)
-def _(expr, langbb):
+def _(expr, langbb, printer):
     if has_integer_args(*expr.args):
         return (('MAX(a,b)', ('(((a) > (b)) ? (a) : (b))')),), {}
     else:
         return (), as_tuple(langbb.get('header-math'))
 
 
 @_lower_macro_math.register(SafeInv)
-def _(expr, langbb):
+def _(expr, langbb, printer):
     try:
-        eps = np.finfo(expr.base.dtype).resolution**2
-    except ValueError:
-        warning(f"dtype not recognized in SafeInv for {expr.base}, assuming float32")
-        eps = np.finfo(np.float32).resolution**2
-    b = Cast('b', dtype=np.float32)
+        dtype = expr.base.dtype
+        eps = np.finfo(dtype).resolution**2
+    except (AttributeError, ValueError):
+        dtype = np.float32
+        eps = np.finfo(dtype).resolution**2
+    b = printer()._print(Cast('b', dtype=dtype))
```

Comment:
can use a blank line before this line

also, instantiating the printer at this point is quite weird. Even more if you think one new instance will be instantiated each time you enter this function. Why don't we create the printer once and for all at the beginning of lowering or wherever (as soon as ...) possible?